plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ContractDescriptionConversionVisitor.kt

```diff
@@ -36,121 +42,139 @@ class ContractDescriptionConversionVisitor(
         returnsEffect: KtReturnsEffectDeclaration<ConeKotlinType, ConeDiagnostic>,
-        data: Unit,
+        data: ContractDescriptionVisitorContext,
     ): Exp {
         val retVar = signature.returnVar.toViper()
         return when (returnsEffect.value) {
-            ConeContractConstantValues.WILDCARD -> BoolLit(true)
+            ConeContractConstantValues.WILDCARD -> BoolLit(true, data.effectSource.asSilverPosition)
             /* NOTE: in a function that has a non-nullable return type, the compiler will not complain if there is an effect like
              * returnsNotNull(). So it is necessary to take care of these cases in order to avoid comparison between non-nullable
              * values and null. In a function that has a non-nullable return type, returnsNotNull() is mapped to true and returns(null)
              * is mapped to false
              */
-            ConeContractConstantValues.NULL -> signature.returnVar.nullCmp(false)
-            ConeContractConstantValues.NOT_NULL -> signature.returnVar.nullCmp(true)
-            ConeContractConstantValues.TRUE -> EqCmp(retVar, BoolLit(true))
-            ConeContractConstantValues.FALSE -> EqCmp(retVar, BoolLit(false))
+            ConeContractConstantValues.NULL -> signature.returnVar.nullCmp(false, data.effectSource)
+            ConeContractConstantValues.NOT_NULL -> signature.returnVar.nullCmp(true, data.effectSource)
+            ConeContractConstantValues.TRUE -> EqCmp(
+                retVar,
+                BoolLit(true, data.effectSource.asSilverPosition),
+                data.effectSource.asSilverPosition
+            )
+            ConeContractConstantValues.FALSE -> EqCmp(
+                retVar,
+                BoolLit(false, data.effectSource.asSilverPosition),
+                data.effectSource.asSilverPosition
+            )
             else -> throw Exception("Unexpected constant: ${returnsEffect.value}")
         }
     }
 
     override fun visitValueParameterReference(
         valueParameterReference: KtValueParameterReference<ConeKotlinType, ConeDiagnostic>,
-        data: Unit,
+        data: ContractDescriptionVisitorContext,
     ): Exp = valueParameterReference.embeddedVar().toViper()
 
     override fun visitIsNullPredicate(
         isNullPredicate: KtIsNullPredicate<ConeKotlinType, ConeDiagnostic>,
-        data: Unit,
+        data: ContractDescriptionVisitorContext,
     ): Exp {
         /* NOTE: useless comparisons like x != null with x non-nullable will compile with just a warning.
          * So it is necessary to take care of these cases in order to avoid comparison between non-nullable
          * values and null. Let x be a non-nullable variable, then x == null is mapped to false and x != null is mapped to true
          */
         val param = isNullPredicate.arg.embeddedVar()
-        return param.nullCmp(isNullPredicate.isNegated)
+        return param.nullCmp(isNullPredicate.isNegated, data.effectSource)
```

Comment:
This (and related lines below) doesn't seem to actually give the accurate position?  You're setting the position of all subnodes to be `data.effectSource`, but they're actually different nodes.  Is there a benefit to having them all refer to the same position?  Shouldn't we be either passing the information correctly, or leaving it blank when we don't know it?